README.md
# Firefox Quantum Themes

The Firefox Quantum Themes for [Visual Studio Code](https://code.visualstudio.com) are inspired by the colours used in Firefox Quantum's DevTools. Using the exact same colours Mozilla use for the chrome and syntax highlighting in the DevTools, I've put together a beautifully dark and light theme. This theme is open-source and will forever be open-source. Anyone is welcome to submit PRs, open issues etc. I would love to hear your feedback! Look below for more information. For Firefox users, this theme will be especially pleasing to use!

## UI Screenshots

![Firefox Quantum Dark](./screenshots/dark-ui.png)

![Firefox Quantum Light](./screenshots/light-ui.png)

For some syntax highlighting examples, see [SCREENSHOTS.md](./SCREENSHOTS.md)

### Supported Languages

By "supported languages", I mean languages where the syntax highlighting doesn't look (too) terrible and is readable.

- Batch/Shell/PowerShell/conf
- C/C++/C# (with their respective extensions)
- Clojure
- CoffeeScript
- CSS/SASS/SCSS/LESS
- Dart
- Dockerfile (with the `jeff-hykin.better-dockerfile-syntax` extension)
- F# (Limited support). Tested with the extension `Ionide for F#: ionide.ionide-fsharp`.
- Go
- GraphQL (with extension `graphql.vscode-graphql-syntax`)
- HTML
- Java
- JavaScript
- JSON/C
- JSX
- Julia
- Lua
- Nix (with `jnoortheen.nix-ide`)
- PHP
- Pug
- Python (with Python extension)
- Ruby
- Rust (via `rust-lang.rust-analyzer`)
- SQL
- TeX/LaTeX/BiBTeX (with extension `james-yu.latex-workshop`)
- TOML (with extension `tamasfe.even-better-toml`)
- TypeScript
- Vue (with extension `vue.volar`)
- XML (with extension `dotjoshjohnson.xml`)
- YAML

## [Changelog](https://github.com/chpxu/vscode-firefox-quantum-themes/blob/master/CHANGELOG.md)

## Theme Colours

For the colours used, please refer to the [Mozilla Developer Network](https://developer.mozilla.org/en-US/docs/Tools/DevToolsColors) for the most up-to-date information.

Here is the quick summary of colours used:

| Dark Hex  | Dark RGBA                 | Light Hex | Light RGBA               |
| --------- | ------------------------- | --------- | ------------------------ |
| `#252c33` | `rgba(37, 44, 51, 1)`     | `#ebeced` | `rgba(235, 236, 237, 1)` |
| `#343c45` | `rgba(52, 60, 69, 1)    ` | `#f0f1f2` | `rgba(240, 241, 242, 1)` |
| `#1d4f73` | `rgba(29, 79, 115, 1)   ` | `#4c9ed9` | `rgba(76, 158, 217, 1)`  |
| `#5c6773` | `rgba(92, 103, 115, 1)  ` | `#747573` | `rgba(116, 117, 115, 1)` |
| `#14171a` | `rgba(17, 19, 21, 1)    ` | `#fcfcfc` | `rgba(252, 252, 252, 1)` |
| `#181d20` | `rgba(24, 29, 32, 1)    ` | `#f7f7f7` | `rgba(247, 247, 247, 1)` |
| `#8fa1b2` | `rgba(143, 161, 178, 1) ` | `#18191a` | `rgba(24, 25, 26, 1)`    |
| `#b6babf` | `rgba(182, 186, 191, 1) ` | `#585959` | `rgba(88, 89, 89, 1)`    |
| `#46afe3` | `rgba(70, 175, 227, 1)  ` | `#0088cc` | `rgba(0, 136, 204, 1)`   |
| `#6b7abb` | `rgba(107, 122, 187, 1) ` | `#5b5fff` | `rgba(91, 95, 255, 1)`   |
| `#df80ff` | `rgba(223, 128, 255, 1) ` | `#b82ee5` | `rgba(184, 46, 229, 1)`  |
| `#eb5368` | `rgba(235, 83, 104, 1)  ` | `#ed2655` | `rgba(237, 38, 85, 1)`   |
| `#d96629` | `rgba(217, 102, 41, 1)  ` | `#f13c00` | `rgba(241, 60, 0, 1)`    |
| `#d99b28` | `rgba(217, 155, 40, 1)  ` | `#d97e00` | `rgba(217, 126, 0, 1)`   |
| `#70bf53` | `rgba(112, 191, 83, 1)  ` | `#2cbb0f` | `rgba(44, 187, 15, 1)`   |
| `#5e88b0` | `rgba(94, 136, 176, 1)  ` | `#0072ab` | `rgba(0, 114, 171, 1)`   |
| `#ffffb4` | `rgba(255, 255, 180, 1)`  | `#ffffb4` | `rgba(255, 255, 180, 1)` |

## Current known differences

See [Differences.md](./DIFFERENCES.md)

## Theme Suggestions

- Have Colourised Bracket pairs enabled.
  - Don't theme it to the same colours as this theme, most colours should work well.
  - Brackets will be much clearer.
- Enable font ligatures if you have a font which supports them.

## Contribution

I currently work on this theme in my spare time. As of writing, it is available in the Marketplace. As a new developer, I am still attempting to work things out, thanks!

If editing and wanting to submit a pull request, please ensure that as much sufficient detail is provided, and briefly state your changes and why, or if meant to fix an issue, reference the problem alongside it.

As of `v2.0.0`, I have changed how the theme files are generated. Upon opening the theme files `Quantum Light.json` and `Quantum Dark.json`, you will see the properties organised terribly. This is because the theme files are generated at the run of a command.

- The files `Quantum Light old.json` and `Quantum Dark old.json` are the legacy theme values (`v1.2.0`). This was for comparison's sake when developing `v2`.

To build and test your modifications, follow these steps:

1. Clone the repository. Have direnv, NodeJS and Yarn installed.
2. Run `direnv allow .`. Then install the dependencies with `yarn install`.
3. You will spend most of your time in the `src` directory. This contains a few TypeScript files:
   - `Types.ts` is a file containing interfaces used to validate variables and functions throughout the project. This will never be edited unless you are wanting to change part of the architecture of the theme.
   - `colours.ts` is the file which contains the colour objects used in the theme.
   - `generator.ts` contains the functions used to generate the theme files.
   - `scopes.ts` contains the attributes to be themed. You will most likely be editing this file a lot to move/add things.
   - `generate.ts` contains all the function callas and post-processing for the theme files. You may be editing this a lot too to call functions with different parameters.
4. In a terminal, run `npm run generate`. This will add theme files to `./themes`. Under the hood, this command runs `ts-node ./src/generate.ts`.
5. Press `F5` (or whatever shortcut you have to debug extensions) and the Extension Host will open up with the new theme files applied. Please ensure you run the generate command after making any changes.

If you wish to lint, you can run `npm run lint`.

### Repo vs Marketplace

The marketplace version always contains the **stable** version of this theme (i.e., whatever is the Published (i.e., exists in Releases) version). The repsitory here on GitHub also contain development pushes (which aren't published on the Marketplace, but tagged on GitHub).

To run these development pushes, follow the Contribution section above.

If you wish to "daily drive" these development pushes, follow the Contribution section, and then:

1. Install the extension from VSCode marketplace regularly
2. Follow Contribution above, in a different directory
3. Copy and paste your generated theme files in `./themes` (in the development version) to the `./themes` in the `extensions` folder of your VSCode install.
4. Restart VSCode
5. If the theme then says a reload is required, DO NOT press it unless you want to reset the extension to stable.
6. Otherwise if you want to go back to stable, uninstall then reinstall the extension

## Thank You's!

This theme was heavily inspired by Firefox DevTools, as well as the [One Dark Pro Port](https://github.com/Binaryify/OneDark-Pro) and VSCode itself! Thanks to them all for the ideas :)

And the following:

- [**@compilenix**](https://github.com/compilenix): First person to show interest and BETA TESTER

- All of you who have even just glanced!

## License

[MIT License](https://github.com/beastdestroyer/vscode-firefox-quantum-themes/blob/master/LICENSE.txt)

Wow, this extension dates back to VSCode 1.26.0!


CHANGELOG.md
# Changelog

Official Releases will be published when there are new versions. This changelog contains dev and stable updates. The stable chnagelog is a merger of all development updates.

## 2.0.0

Wow, it finally happened after 1 year.

### Development

- Development has moved to TypeScript! See [README](./README.md) for more details. Theme files are now generated at the run of a command rather than manually modifying JSON files.
  - Created function logic to generate `tokenColors` and `colors`.
- ESLint to lint the TS files. Currently default settings (`recommended`).
- Added `direnv` and `nix` files for development.
- Alphabetical sort of all entries (and consequently means every scope in one tokenColor is sorted by category)

### Updates

- Updated [README.md](./README.md)
- `DIFFERENCES.md` keeps track of known differences with current devtools alongside any reasoning/ways forward
- Added `editor.rangeHighlightBackground`
- Added support for `.nix` files
- All comments are now the colours of `comment`.
- Added support for `dockerfile` via the extension `jeff-hykin.better-dockerfile-syntax`.
- Added limited support for `F#`. Tested with the extension `Ionide for F#: ionide.ionide-fsharp`.
- Added support for Rust (via `rust-lang.rust-analyzer`) keywords and namespaces.
- Improved support for other various inbuilt languages. Everything else is mostly the same.
- Improved theming for the status bar: debuggging, status items etc.
- No white text on light theme.
- Buttons and badges match active theme.
- Updated screenshots.

## 1.2.0

### Colour Theming

- `activityBar.activeBorder` is now blue as it is in DevTools

### Syntax Highlighting

- PHP
  - Changed type highlighting in PHP to be yellow #43 (light orange in light theme)
  - Embedded SQL follows the changes made in "SQL"
- C#
  - Changed type and delegate highlighting to be yellow
  - Changed fields and event variables to be blue
- YAML:
  - Types yellow (light orange in light theme)
  - Dates are now green
- SQL:
  - Params/variables are blue
  - Bracketed text is now blue-grey
  - DML keywords are light orange
  - Other keywords are pink except aliases - they remain green.
- Clojure
  - Meta symbols are orange
  - Keyword constants are blue
- Dart
  - Classes are now yellow (light orange in light theme)
  - Regular variables are default text colour for this theme. They do not hae have a TextMate scope and thus I cannot make them blue as desired

### Other

- Screenshots added: PHP, SQL, C#, YAML, Dart
- README updated

The next beta (currently) will focus on cleaning up the repository, highlighting more languages and focusing on the minimap/decorations. This version is NOT on the marketplace.

## 1.1.9

- Made selection highlight in dark theme have better accessibility contrast
- Changed some other things such as bracket highlighting slightly
- Cursor now has a yellow tint

## 1.1.8

- Made `storage.type.annotations` clearer to see in Light
- Made user-defined generic functions highlighted on languages such as Python
- Some misc back-end stuff

## 1.1.5

- In-built classes for languages should now be pink (in accordance to Firefox DevTool/Scratchpad)
- JS/JSON Property keys now in blue

- Polish to repo

## 1.1.3

- Variables now blue, functions now green
- Support classes, etc are now purple

## 1.1.1

- Fixed formatting of SCREENSHOTS.md

## 1.1.0

Here lands a biggish update, including improvements to the theme to catch up to the Firefox DevTools, and improvements to repo structure.

- Features:

  - Support for basically all the other default languages
  - Clean(er) repository
    - Removal of test files
    - Better screenshots
    - Improved Wiki

- Tweaks
  - Maintaining parity with Firefox DevTools [#28](https://github.com/beastdestroyer/vscode-firefox-quantum-themes/issues/28)
    - Variable colours now blue instead of red
    - Property colours now green
    - Numbers and interpolation now a more vibrant blue
    - Support types and CSS Pseudos now a lighter/darker grey for dark/light themes respectively
  - Operators now coloured the same as types (light grey) to maintain consistency with DevTools, and to stop the amount of blue!
  - Symbols in Clojure changed to green
  - CONSISTENCY Background colour changed on tabs to match DevTools, Light
  - CONSISTENCY Tab border changed, Light
  - CONSISTENCY Selection now transparent to match DevTools and for better visibility

## 1.0.4

- Improved Language Support for:

  - Perl/Perl6
  - PHP
  - Python
  - PowerShell
  - log
  - R
  - Razor (cshtml) \[there are a few glitches with highlighting, but it appears to be with VSCode itself not the theme after further investigation]

- Select in Dropdowns now grey again! (Quantum Light)
- Theme title is not the same as name of extension [#25](https://github.com/beastdestroyer/vscode-firefox-quantum-themes/issues/25)
- Yellow not visible in Quantum Light [#22](https://github.com/beastdestroyer/vscode-firefox-quantum-themes/issues/22)

## 1.0.2

- Fix metas #16 #21 @compilenix
- Made cursor brighter #17 @compilenix
- Brighter text, turns out it matches Firefox :) #18 #19 #20 @compilenix

## 1.0.0

- Initial release


src/colours.ts
// global variables for colours and attributes to theme
// These have the same name as they are listed here:
// https://firefox-source-docs.mozilla.org/devtools-user/devtoolscolors/index.html
// but without the "--theme" and all hyphens replaced with underscores
// All colours in the form of an array with [dark colour, light colour]
// https://stackoverflow.com/questions/13612006/get-object-property-name-as-a-string
import { Colour, Colours } from "./Types";

export function key_name<T extends object>(
  obj: T,
  expression: (x: { [Property in keyof T]: () => string }) => () => string
): string {
  const res: { [Property in keyof T]: () => string } = {} as {
    [Property in keyof T]: () => string;
  };
  Object.keys(obj).map((k) => (res[k as keyof T] = () => k));
  return expression(res)();
}

// VSCode Scopes
// https://code.visualstudio.com/api/references/theme-color

const chrome_colours: Colours = {
  // CHROME COLOURS
  tab_toolbar_background: ["#252c33", "#ebeced"],
  toolbar_background: ["#343c45", "#f0f1f2"],
  selection_background: ["#1d4f73", "#4c9ed9"],
  selection_color: ["#f5f7fa", "#f5f7fa"],
  splitter_color: ["#000000", "#aaaaaa"],
  comment: ["#5c6773", "#747573"],
};

// CONTENT COLOURS
const ui_colours: Colours = {
  body_background: ["#14171a", "#fcfcfc"],
  sidebar_background: ["#181d20", "#f7f7f7"],
  contrast_background: ["#b28025", "#e6b064"],
};

const text_colours: Colours = {
  body_color: ["#8fa1b2", "#18191a"],
  body_color_alt: ["#b6babf", "#585959"],
  content_color1: ["#a9bacb", "#292e33"],
  content_color2: ["#8fa1b2", "#8fa1b2"],
  content_color3: ["#667380", "#667380"],
};

export const highlight_colors: Colours = {
  highlight_blue: ["#46afe3", "#0088cc"],
  highlight_purple: ["#6b7abb", "#5b5fff"],
  highlight_pink: ["#df80ff", "#b82ee5"],
  highlight_red: ["#eb5368", "#ed2655"],
  highlight_orange: ["#d96629", "#f13c00"],
  highlight_lightorange: ["#d99b28", "#d97e00"],
  highlight_green: ["#70bf53", "#2cbb0f"],
  highlight_bluegrey: ["#5e88b0", "#0072ab"],
  highlight_yellow: ["#ffffb4", "#d97e00"],
};
// === Create colors object to loop through and do highlighting
// const coloursAttr: Colour[] = [];
export const all_colours: Colours = {
  ...chrome_colours,
  ...ui_colours,
  ...text_colours,
  ...highlight_colors,
};
